File: credit/credit.c
#include <stdio.h>
#include <cs50.h>

void print_credit_card_brand(long long ccn);
bool checksum(long long ccn);
int find_length(long long n);
bool check_validity(long long credit_card_number);

int main(void)
{
    long long credit_card_number;
    do
    {
        credit_card_number = get_long_long("Number: ");
    }
    while (credit_card_number < 0);

    if(check_validity(credit_card_number))
        print_credit_card_brand(credit_card_number);
    else
        printf("INVALID\n");
}

bool check_validity(long long credit_card_number)
{
    int len = find_length(credit_card_number);
    return (len == 13 || len == 15 || len == 16) && checksum(credit_card_number);
}

int find_length(long long n)
{
    int len;
    for (len = 0;n != 0; n /= 10, len++);
    return len;
}

bool checksum(long long ccn)
{
    int sum = 0;
    for(int i = 0; ccn != 0; ccn /= 10, i++)
    {
        if (i % 2 == 0)
        {
            sum += ccn % 10;
        }
        else
        {
            int digit = 2 * (ccn % 10);
            sum += digit / 10 + digit % 10;
        }
    }
    return (sum % 10) == 0;
}

void print_credit_card_brand(long long ccn)
{
    if ((ccn >= 34e13 && ccn < 35e13) || (ccn >= 37e13 && ccn < 38e13))
        printf("AMEX\n");
    else if (ccn >= 51e14 && ccn < 56e14)
        printf("MASTERCARD\n");
    else if ((ccn >= 4e12 && ccn < 5e12) || (ccn >= 4e15 && ccn < 5e15))
        printf("VISA\n");
    else
        printf("INVALID\n");
}

File: sentimental-credit/credit.py
from cs50 import get_string

# Get the number to the variable and pass it to an unique function that checks and print it
def main():
    card_number = get_string("Card Number: ")
    print_card_flag(card_number)

# Check the functions that fits in and then print it flag
def print_card_flag(card_number):
    if (check_sum(card_number)) == False:
        print("INVALID")
    elif (check_amex(card_number)) == True:
        print("AMEX")
    elif (check_mastercard(card_number)) == True:
        print("MASTERCARD")
    elif (check_visa(card_number)) == True:
        print("VISA")

def check_sum(card_number):
    # initialise variables
    sum = 0
    length = len(card_number)

    # sum_of_double(i) = sum of digits in 2i. defined for i in range(10)
    sum_of_double = [0, 2, 4, 6, 8, 1, 3, 5, 7, 9]

    # loop through card_number and add the digits contribution to checksum
    for i in range(length):
        ith_digit_from_right = int(card_number[length-1-i])

        # if i is even, then we are in odd position from right, so just add ith digit
        if i % 2 == 0:
            sum += ith_digit_from_right

        # if i is odd, we are at even position from right, so add double of ith digit
        else:
            sum += sum_of_double[ith_digit_from_right]

    # if final digit in sum is 0, then passes checksum test
    if sum % 10 == 0:
        return True
    else:
        return False

def check_amex(card_number):
    # if card_number satisfies amex conditions, return True otherwise false
    if len(card_number) == 15 and int(card_number[:2]) in [34,37]:
        return True
    else:
        return False

def check_mastercard(card_number):
        # if card_number satisfies amex conditions, return True otherwise false
    if len(card_number) == 16 and int(card_number[:2]) in [51, 51, 53, 54, 55]:
        return True
    else:
        return False

def check_visa(card_number):
        # if card_number satisfies amex conditions, return True otherwise false
    if len(card_number) in [13,16] and int(card_number[0]) == 4:
        return True
    else:
        return False

# Run the programming
main()